docs/book/src/reference/cli-plugins.md

```diff
@@ -1,16 +1,262 @@
+      &helmv1.Plugin{},    // "helm.example.com/v1.0"
+      &ansiblev1.Plugin{}, // "ansible.example.com/v1.0"
+    ),
+    cli.WithDefaultPlugins(
+      &golangv3.Plugin{},  // "go.example.com/v2.0", default
+    ),
+  )
+  if err != nil {
+    log.Fatal(err)
+  }
+  if err := c.Run(); err != nil {
+    log.Fatal(err)
+  }
+}
+```
+
+This program can then be built and run in the following ways:
+
+Default behavior:
+
+```sh
+# Initialize a project with the default Init plugin, "go.example.com/v2.0". This key is automatically written to
+# a PROJECT config file.
+$ controller-builder init
+# Create an API and webhook with "go.example.com/v2.0" CreateAPI and CreateWebhook plugin methods. This key was read
+# from the config file.
+$ controller-builder create api [flags]
+$ controller-builder create webhook [flags]
+```
+
+Selecting a plugin using `--plugins`:
+
+```sh
+# Initialize a project with the "ansible.example.com/v2.0" Init plugin. Like above, this key is written to a config file.
+$ controller-builder init
+# Create an API and webhook with "ansible.example.com/v2.0" CreateAPI and CreateWebhook plugin methods. This key was read
+# from the config file.
+$ controller-builder create api [flags]
+$ controller-builder create webhook [flags]
+```
+
+## Plugin naming
+
+Plugin names must be DNS1123 labels and should be fully qualified, i.e. they have a suffix like
+`.example.com`. For example, the base Go scaffold used with `kubebuilder` commands has name `go.kubebuilder.io`.
+Qualified names prevent conflicts between plugin names; both `go.kubebuilder.io` and `go.example.com` can both scaffold
+Go code and can be specified by a user.
+
+## Plugin versioning
+
+Plugin versions are valid "tolerant" semver, in that they can optionally contain a "v" prefix
+and patch version is not used nor allowed. Changes made to plugins may require a version increase depending on whether
+they break the current version or not. See the [contribution guide][contrib-guide-ex] for a version change example.
+
+### Breaking changes
+
+Situations that result in a breaking change:
+* Deleting a default scaffold file.
+* Modifying a scaffold file that will be modified by other commands.
+* Changing the path to a default scaffold file.
+
+Simply put, any change that will break the expected behavior of a project built with the previous _minor_ plugin version
+is a breaking change.
+
+Breaking changes require an increase in a plugin's _major_ version.
+
+### Non-breaking changes
+
+There are two types of non-breaking changes: those requiring a _minor_ version increase, and those that do not
+require a version be changed.
+
+Situations that require a _minor_ version bump:
+* Adding a new file that other commands do not assume exists.
+* Modifying a scaffold file in a way that does not affect other commands.
+
+Situations that do not require a version be changed:
+* Typo fixes.
+* Cosmetic changes that do not affect program behavior.
```

Comment:
I am not just ok with this ^. Both examples may be or not breaking changes. So, I would suggest just remove it.

Reply:
We need to define when a version does not need to be changed. Do you have alternative suggestions?

Reply:
So, I understand that; 

- Any change which will result in a scaffold change made by the plugins requires a new mInor/major plugin bump
- Any change that affects the behaviour of the plugin also shows requires a bump right?

However, if we do a change that has no relation with the code scaffolded by the plugin or its behaviours do not require a plugin bump. E.g 
- in the edit/autocomplete commands that are not using a plugin
- or in the books/ci/ or indeed in the tests.

Reply:
Those things do not relate to plugins though. How does a typo fix cause any issue when scaffolding? Once again, #1498 is not a good example because that isn't a typo _and_ changes program behavior.

Reply:
I think I got what you mean here .You can close this as solved